streaming/src/main/scala/org/apache/spark/streaming/StreamingContext.scala

```diff
@@ -714,12 +714,20 @@ class StreamingContext private[streaming] (
           // interrupted. See SPARK-12001 for more details. Because the body of this case can be
           // executed twice in the case of a partial stop, all methods called here need to be
           // idempotent.
-          scheduler.stop(stopGracefully)
```

Comment:
I think one of the primary goal of this JIRA is to allow partial clean-up and retry on `stop()` calls.

In this specific code path, it is already written in a way to allow for retry by setting the state to `STOPPED` only almost at the end on [line 728](https://github.com/apache/spark/pull/10807/files#diff-8a7f0e3f26c15ba484e6312c3caf033dL728) in the original code. 

`tryLogNonFatalError` swallows and logs "non-fatal" exception, and with that added, despite any non-critical error thrown it could reach the line `state = STOPPED`. For instance, if `env.metricsSystem.removeSource()` throws then it will continue on and setting `state` to `STOPPED`, at which point the caller cannot get back to the same code to retry cleanup because of the `state` match case [above](https://github.com/apache/spark/pull/10807/files#diff-8a7f0e3f26c15ba484e6312c3caf033dR707).

Is that what we want?